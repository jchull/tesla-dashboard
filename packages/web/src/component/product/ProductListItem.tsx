import React from 'react'
import numbro from 'numbro'
import { BatteryLevelIcon } from '../common/BatteryLevelIcon'
import moment from 'moment'
import { ProductListItemState } from '@teslapp/web/src/store/types/state'

export const ProductListItem: React.FC<ProductListItemState> = (
  props: ProductListItemState
) => {
  const timeToFull = moment
    .duration(props.product.time_to_full_charge || 0, 'hours')
    .humanize()
  return (
    <div
      className="list-item product red_pinwheel18"
      onClick={() => props.handleSelection(props.product._id)}
    >
      <div className="row">
        <h2 className="name">{props.product.display_name}</h2>
        <div className="battery_level end">
          <span>{props.product.battery_level}%</span>
          <BatteryLevelIcon
            batteryLevel={props.product.battery_level}
            batteryRange={props.product.battery_range}
            chargingState={props.product.charging_state}
            chargeLimit={90}
            width={40}
          />
          <span>{numbro(props.product.battery_range)
            .format('0,0.0')} mi</span>
        </div>
      </div>
      <div className="row">
        <div className="odometer start">
          {numbro(props.product.odometer)
            .format('0,0.0')} miles
        </div>
      </div>

      <div className="row">
        <div className="status start">
          as of {moment(props.product.timestamp).fromNow()}
        </div>
       <div className="status end">
         {props.product.charging_state === 'Charging'
           ? `${timeToFull} remaining`
           : props.product.state}
        </div>

      </div>
    </div>
  )
}
