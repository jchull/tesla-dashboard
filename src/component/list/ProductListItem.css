.product {
    font-weight: 700;
    box-shadow: 1px 1px 1px #aaaaaa;
    height: 100px;
    width: 300px;
    color: #444444;
}

.product:hover {
    cursor: pointer;
    box-shadow: 1px 1px 1px #777777;
}

.product:active {
    transform: translate(1px, 1px);
}

.products .list-item .name {
    padding-top: 2px;
}

.battery_level {
    padding-top: 0;
}

.battery_level i {
    line-height: 30px;
}

.battery_level i.charging {
    transform: rotate(90deg);
    padding-left: 4px;
    width: 40px;
    font-size: 180%;
}

.battery_level i.charging {
    color: #00b32a;
}
