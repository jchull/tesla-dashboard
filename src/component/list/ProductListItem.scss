.product.list-item {
  border: none;
  display: inline-block;
  margin: 6px;
  font-weight: 400;
  box-shadow: 1px 1px 3px rgba(102, 102, 102, 0.2);
  height: 100px;
  width: 360px;
  color: #222222;
  transition-duration: 300ms;

  .name {
    padding: 0 2px;
  }

  .battery_level span {
    font-size: 21px;
    line-height: 21px;
    vertical-align: top;
    padding: 0 0 0 2px;
  }
}

.product.list-item:hover {
  transition-duration: 150ms;
  box-shadow: 1px 2px 4px rgba(102, 102, 102, 0.2);
  transform: translate3d(1px, -1px, 3px);
}

.product.list-item:active {
  transform: translate(-1px, 1px);
}


.product.list-item.red_pinwheel18 {
  background: url("/images/tm3_red_pinwheel18.png") #f7f7f7;
  background-size: 160px;
  background-repeat: no-repeat;
  background-position: 215px 25px;
}

.better {
  color: #199f07;
}

.worse {
  color: rgb(227, 25, 55);
}
