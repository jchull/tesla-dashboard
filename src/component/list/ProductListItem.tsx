import React from 'react';
import './ProductListItem.scss';
import numbro from 'numbro';
import {BatteryLevelIcon} from '../common/BatteryLevelIcon';
import moment from 'moment';
import {ProductListItemState} from '../../type/state';


export const ProductListItem: React.FC<ProductListItemState> = (props: ProductListItemState) => {

  const timeToFull = moment.duration(props.product.time_to_full_charge || 0, 'hours')
                           .humanize();
  return (
      <div className="list-item product red_pinwheel18"
           onClick={() => props.handleSelection(props.product)}>
        <div className="row">
          <h2 className="name">{props.product.display_name}</h2>
          <div className="battery_level end">
            <span>{props.product.battery_level}%</span>
            <BatteryLevelIcon battery_level={props.product.battery_level}
                              battery_range={props.product.battery_range}
                              charging_state={props.product.charging_state}
                              charge_limit={90}
                              width={40}/>
            <span>{numbro(props.product.battery_range)
                .format('0,0.0')} mi</span>
          </div>
        </div>
        <div className="row">
          <div className="odometer start">{numbro(props.product.odometer)
              .format('0,0.0')} miles
          </div>


        </div>

        <div className="row">
          <div className="status start">
            {
              props.product.charging_state === 'Charging' ?
                  `${timeToFull} remaining`
                  :
                  props.product.state
            }
          </div>


        </div>

      </div>
  );
};

