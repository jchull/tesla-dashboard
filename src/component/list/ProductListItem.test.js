import React                      from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import pretty                     from 'pretty';


import { ProductListItem } from './ProductListItem';

let wrapper;

beforeEach(() => {
});

afterEach(() => {
    if(wrapper) {
        if(wrapper.unmount) {
            wrapper.unmount();
        }
        if(wrapper.container) {
            wrapper.container = null;
        }
    }
});


it('renders product list item', () => {
    const product = {
        '_id': 'aaaaaaaaa',
        'vin': 'abc',
        'display_name': 'Display Name',
        'option_codes': '',
        'color': 'RedMulticoat',
        'state': 'Driving',
        'id_s': '123456789',
        'calendar_enabled': true,
        'api_version': 6,
        'battery_level': 60,
        'car_type': 'model3',
        'odometer': 123.45,
        'timestamp': 1566408896502,
        'battery_range': 182.91,
        'charging_state': 'Disconnected',
        'time_to_full_charge': 0,
        'charge_limit_soc': 83,
        'charge_limit_soc_min': 50,
        'last_session_id': 'aadzfsfdsfa'
    };
    const handler = (selectedProduct) => console.log(selectedProduct.display_name);

    act(() => {
        wrapper = render(<ProductListItem selected={false}
            product={product}
            handleSelection={handler}/>);
    });
    expect(pretty(wrapper.container.innerHTML))
        .toMatchSnapshot();
});
