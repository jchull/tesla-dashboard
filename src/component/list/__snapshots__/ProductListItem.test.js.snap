// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders product list item 1`] = `
"<div class=\\"list-item product red_pinwheel18\\">
  <div class=\\"row\\">
    <h2 class=\\"name\\">Display Name</h2>
    <div class=\\"battery_level end\\"><span>60%</span>
      <div class=\\"battery-icon\\" style=\\"width: 40px;\\"><svg viewBox=\\"0 0 40 20\\"></svg></div><span>182.9 mi</span></div>
  </div>
  <div class=\\"row\\">
    <div class=\\"odometer start\\">123.5 miles</div>
  </div>
  <div class=\\"row\\">
    <div class=\\"status start\\">Driving</div>
  </div>
</div>"
`;
